File: app/src/main/java/me/zhanghai/android/files/filelist/FileListAdapter.java
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filelist;

import android.content.Context;
import android.graphics.drawable.Drawable;
import android.provider.DocumentsContract;
import android.text.TextUtils;
import android.view.Menu;
import android.view.View;
import android.view.ViewGroup;
import android.widget.ImageButton;
import android.widget.ImageView;
import android.widget.TextView;

import com.bumptech.glide.signature.ObjectKey;
import com.simplecityapps.recyclerview_fastscroll.views.FastScrollRecyclerView;

import java.util.Comparator;
import java.util.HashMap;
import java.util.HashSet;
import java.util.Iterator;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Locale;
import java.util.Map;
import java.util.Objects;
import java.util.Set;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.content.res.AppCompatResources;
import androidx.appcompat.widget.PopupMenu;
import androidx.fragment.app.Fragment;
import androidx.recyclerview.widget.RecyclerView;
import androidx.recyclerview.widget.SortedList;
import androidx.recyclerview.widget.SortedListAdapterCallback;
import butterknife.BindView;
import butterknife.ButterKnife;
import java8.nio.file.Path;
import java8.nio.file.attribute.BasicFileAttributes;
import me.zhanghai.android.files.R;
import me.zhanghai.android.files.compat.StringCompat;
import me.zhanghai.android.files.file.FormatUtils;
import me.zhanghai.android.files.file.MimeTypes;
import me.zhanghai.android.files.glide.GlideApp;
import me.zhanghai.android.files.glide.IgnoreErrorDrawableImageViewTarget;
import me.zhanghai.android.files.provider.archive.ArchiveFileSystemProvider;
import me.zhanghai.android.files.provider.document.DocumentFileAttributes;
import me.zhanghai.android.files.provider.document.DocumentFileSystemProvider;
import me.zhanghai.android.files.provider.linux.LinuxFileSystemProvider;
import me.zhanghai.android.files.settings.SettingsLiveDatas;
import me.zhanghai.android.files.ui.AnimatedSortedListAdapter;
import me.zhanghai.android.files.ui.CheckableFrameLayout;
import me.zhanghai.android.files.util.ViewUtils;

public class FileListAdapter extends AnimatedSortedListAdapter<FileItem, FileListAdapter.ViewHolder>
        implements FastScrollRecyclerView.SectionedAdapter {

    private static final Object PAYLOAD_STATE_CHANGED = new Object();

    private boolean mSearching;
    @NonNull
    private Comparator<FileItem> mComparator;
    @NonNull
    private final SortedList.Callback<FileItem> mCallback =
            new SortedListAdapterCallback<FileItem>(this) {
                @Override
                public int compare(FileItem file1, FileItem file2) {
                    if (mSearching) {
                        return ((SearchFileItem) file1).compareTo((SearchFileItem) file2);
                    } else {
                        return mComparator.compare(file1, file2);
                    }
                }
                @Override
                public boolean areItemsTheSame(FileItem oldItem, FileItem newItem) {
                    return Objects.equals(oldItem, newItem);
                }
                @Override
                public boolean areContentsTheSame(FileItem oldItem, FileItem newItem) {
                    return FileItem.contentEquals(oldItem, newItem);
                }
            };

    @Nullable
    private PickOptions mPickOptions;

    @NonNull
    private final LinkedHashSet<FileItem> mSelectedFiles = new LinkedHashSet<>();
    @NonNull
    private final Map<FileItem, Integer> mFilePositionMap = new HashMap<>();

    @NonNull
    private Fragment mFragment;
    @NonNull
    private Listener mListener;

    public FileListAdapter(@NonNull Fragment fragment, @NonNull Listener listener) {
        init(FileItem.class, mCallback);

        mFragment = fragment;
        mListener = listener;
    }

    public void setComparator(@NonNull Comparator<FileItem> comparator) {
        mComparator = comparator;
        if (!mSearching) {
            refresh();
            rebuildFilePositionMap();
        }
    }

    public void setPickOptions(@Nullable PickOptions pickOptions) {
        mPickOptions = pickOptions;
        notifyItemRangeChanged(0, getItemCount(), PAYLOAD_STATE_CHANGED);
    }

    public void replaceSelectedFiles(@NonNull LinkedHashSet<FileItem> files) {
        Set<FileItem> changedFiles = new HashSet<>();
        for (Iterator<FileItem> iterator = mSelectedFiles.iterator(); iterator.hasNext(); ) {
            FileItem file = iterator.next();
            if (!files.contains(file)) {
                iterator.remove();
                changedFiles.add(file);
            }
        }
        for (FileItem file : files) {
            if (!mSelectedFiles.contains(file)) {
                mSelectedFiles.add(file);
                changedFiles.add(file);
            }
        }
        for (FileItem file : changedFiles) {
            Integer position = mFilePositionMap.get(file);
            if (position != null) {
                notifyItemChanged(position, PAYLOAD_STATE_CHANGED);
            }
        }
    }

    private void selectFile(@NonNull FileItem file) {
        if (!isFileSelectable(file)) {
            return;
        }
        if (mPickOptions != null && !mPickOptions.allowMultiple) {
            mListener.clearSelectedFiles();
        }
        mListener.selectFile(file, !mSelectedFiles.contains(file));
    }

    public void selectAllFiles() {
        LinkedHashSet<FileItem> files = new LinkedHashSet<>();
        for (int i = 0, count = getItemCount(); i < count; ++i) {
            FileItem file = getItem(i);
            if (isFileSelectable(file)) {
                files.add(file);
            }
        }
        mListener.selectFiles(files, true);
    }

    private boolean isFileSelectable(@NonNull FileItem file) {
        if (mPickOptions != null) {
            return !file.getAttributes().isDirectory() && mPickOptions.mimeTypeMatches(
                    file.getMimeType());
        }
        return true;
    }

    @Override
    public void clear() {
        super.clear();

        rebuildFilePositionMap();
    }

    /**
     * @deprecated Use {@link #replace2(List, boolean)} instead.
     */
    @Override
    public void replace(@NonNull List<FileItem> list, boolean clear) {
        throw new UnsupportedOperationException();
    }

    public void replace2(@NonNull List<FileItem> list, boolean searching) {

        boolean clear = mSearching != searching;
        mSearching = searching;

        super.replace(list, clear);

        rebuildFilePositionMap();
    }

    private void rebuildFilePositionMap() {
        mFilePositionMap.clear();
        for (int i = 0, count = getItemCount(); i < count; ++i) {
            FileItem file = getItem(i);
            mFilePositionMap.put(file, i);
        }
    }

    @NonNull
    @Override
    public ViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        ViewHolder holder = new ViewHolder(ViewUtils.inflate(R.layout.file_item, parent));
        holder.itemLayout.setBackground(AppCompatResources.getDrawable(
                holder.itemLayout.getContext(), R.drawable.checkable_item_background));
        holder.menu = new PopupMenu(holder.menuButton.getContext(), holder.menuButton);
        holder.menu.inflate(R.menu.file_item);
        holder.menuButton.setOnClickListener(view -> holder.menu.show());
        holder.menuButton.setOnTouchListener(holder.menu.getDragToOpenListener());
        return holder;
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position) {
        throw new UnsupportedOperationException();
    }

    @Override
    public void onBindViewHolder(@NonNull ViewHolder holder, int position,
                                 @NonNull List<Object> payloads) {
        FileItem file = getItem(position);
        boolean isDirectory = file.getAttributes().isDirectory();
        boolean enabled = isFileSelectable(file) || isDirectory;
        holder.itemLayout.setEnabled(enabled);
        holder.iconLayout.setEnabled(enabled);
        holder.menuButton.setEnabled(enabled);
        Menu menu = holder.menu.getMenu();
        Path path = file.getPath();
        boolean hasPickOptions = mPickOptions != null;
        boolean isReadOnly = path.getFileSystem().isReadOnly();
        menu.findItem(R.id.action_cut).setVisible(!hasPickOptions && !isReadOnly);
        menu.findItem(R.id.action_copy).setVisible(!hasPickOptions);
        boolean checked = mSelectedFiles.contains(file);
        holder.itemLayout.setChecked(checked);
        if (!payloads.isEmpty()) {
            return;
        }
        bindViewHolderAnimation(holder);
        holder.itemLayout.setOnClickListener(view -> {
            if (mSelectedFiles.isEmpty()) {
                mListener.openFile(file);
            } else {
                selectFile(file);
            }
        });
        holder.itemLayout.setOnLongClickListener(view -> {
            if (mSelectedFiles.isEmpty()) {
                selectFile(file);
            } else {
                mListener.openFile(file);
            }
            return true;
        });
        holder.iconLayout.setOnClickListener(view -> selectFile(file));
        String mimeType = file.getMimeType();
        Drawable icon = AppCompatResources.getDrawable(holder.iconImage.getContext(),
                MimeTypes.getIconRes(mimeType));
        BasicFileAttributes attributes = file.getAttributes();
        if (supportsThumbnail(file)) {
            GlideApp.with(mFragment)
                    .load(path)
                    .signature(new ObjectKey(attributes.lastModifiedTime()))
                    .placeholder(icon)
                    .into(new IgnoreErrorDrawableImageViewTarget(holder.iconImage));
        } else {
            GlideApp.with(mFragment)
                    .clear(holder.iconImage);
            holder.iconImage.setImageDrawable(icon);
        }
        Integer badgeIconRes;
        if (file.getAttributesNoFollowLinks().isSymbolicLink()) {
            badgeIconRes = file.isSymbolicLinkBroken() ? R.drawable.error_badge_icon_18dp
                    : R.drawable.symbolic_link_badge_icon_18dp;
        } else {
            badgeIconRes = null;
        }
        boolean hasBadge = badgeIconRes != null;
        ViewUtils.setVisibleOrGone(holder.badgeImage, hasBadge);
        if (hasBadge) {
            holder.badgeImage.setImageResource(badgeIconRes);
        }
        holder.nameText.setText(FileUtils.getName(file));
        String description;
        if (isDirectory) {
            description = null;
        } else {
            Context context = holder.descriptionText.getContext();
            String descriptionSeparator = context.getString(
                    R.string.file_item_description_separator);
            String lastModificationTime = FormatUtils.formatShortTime(
                    attributes.lastModifiedTime().toInstant(), context);
            String size = FormatUtils.formatHumanReadableSize(attributes.size(), context);
            description = StringCompat.join(descriptionSeparator, lastModificationTime, size);
        }
        holder.descriptionText.setText(description);
        boolean isArchivePath = ArchiveFileSystemProvider.isArchivePath(path);
        menu.findItem(R.id.action_copy).setTitle(isArchivePath ? R.string.file_item_action_extract
                : R.string.copy);
        menu.findItem(R.id.action_delete).setVisible(!isReadOnly);
        menu.findItem(R.id.action_rename).setVisible(!isReadOnly);
        boolean isArchiveFile = FileUtils.isArchiveFile(file);
        menu.findItem(R.id.action_extract).setVisible(isArchiveFile);
        holder.menu.setOnMenuItemClickListener(item -> {
            switch (item.getItemId()) {
                case R.id.action_open_as:
                    mListener.showOpenFileAsDialog(file);
                    return true;
                case R.id.action_cut:
                    mListener.cutFile(file);
                    return true;
                case R.id.action_copy:
                    mListener.copyFile(file);
                    return true;
                case R.id.action_delete:
                    mListener.confirmDeleteFile(file);
                    return true;
                case R.id.action_rename:
                    mListener.showRenameFileDialog(file);
                    return true;
                case R.id.action_extract:
                    mListener.extractFile(file);
                    return true;
                case R.id.action_archive:
                    mListener.showCreateArchiveDialog(file);
                    return true;
                case R.id.action_send:
                    mListener.sendFile(file);
                    return true;
                case R.id.action_copy_path:
                    mListener.copyPath(file);
                    return true;
                case R.id.action_properties:
                    mListener.showPropertiesDialog(file);
                    return true;
                default:
                    return false;
            }
        });
    }

    private static boolean supportsThumbnail(@NonNull FileItem file) {
        Path path = file.getPath();
        if (LinuxFileSystemProvider.isLinuxPath(path)) {
            return MimeTypes.supportsThumbnail(file.getMimeType());
        } else if (DocumentFileSystemProvider.isDocumentPath(path)) {
            DocumentFileAttributes attributes = (DocumentFileAttributes) file.getAttributes();
            return (attributes.getFlags() & DocumentsContract.Document.FLAG_SUPPORTS_THUMBNAIL)
                    == DocumentsContract.Document.FLAG_SUPPORTS_THUMBNAIL;
        } else {
            // TODO: Allow other providers as well - but might be resource consuming.
            return false;
        }
    }

    @NonNull
    @Override
    public String getSectionName(int position) {
        FileItem file = getItem(position);
        String name = FileUtils.getName(file);
        if (TextUtils.isEmpty(name)) {
            return "";
        }
        return name.substring(0, 1).toUpperCase(Locale.getDefault());
    }

    @Override
    protected boolean isAnimationEnabled() {
        return SettingsLiveDatas.FILE_LIST_ANIMATION.getValue();
    }

    public interface Listener {
        void clearSelectedFiles();
        void selectFile(@NonNull FileItem file, boolean selected);
        void selectFiles(@NonNull LinkedHashSet<FileItem> files, boolean selected);
        void openFile(@NonNull FileItem file);
        void showOpenFileAsDialog(@NonNull FileItem file);
        void cutFile(@NonNull FileItem file);
        void copyFile(@NonNull FileItem file);
        void confirmDeleteFile(@NonNull FileItem file);
        void showRenameFileDialog(@NonNull FileItem file);
        void extractFile(@NonNull FileItem file);
        void showCreateArchiveDialog(@NonNull FileItem file);
        void sendFile(@NonNull FileItem file);
        void copyPath(@NonNull FileItem file);
        void showPropertiesDialog(@NonNull FileItem file);
    }

    static class ViewHolder extends RecyclerView.ViewHolder {

        @BindView(R.id.item)
        public CheckableFrameLayout itemLayout;
        @BindView(R.id.icon_layout)
        public ViewGroup iconLayout;
        @BindView(R.id.icon)
        public ImageView iconImage;
        @BindView(R.id.badge)
        public ImageView badgeImage;
        @BindView(R.id.name)
        public TextView nameText;
        @BindView(R.id.description)
        public TextView descriptionText;
        @BindView(R.id.menu)
        public ImageButton menuButton;

        @NonNull
        public PopupMenu menu;

        public ViewHolder(@NonNull View itemView) {
            super(itemView);

            ButterKnife.bind(this, itemView);
        }
    }
}

File: app/src/main/java/me/zhanghai/android/files/filelist/FileListAdapter.kt
/*
 * Copyright (c) 2018 Hai Zhang <dreaming.in.code.zh@gmail.com>
 * All Rights Reserved.
 */

package me.zhanghai.android.files.filelist

import android.text.TextUtils
import android.view.ViewGroup
import androidx.appcompat.widget.PopupMenu
import androidx.core.view.isVisible
import androidx.recyclerview.widget.DiffUtil
import androidx.recyclerview.widget.RecyclerView
import coil.clear
import coil.loadAny
import java8.nio.file.Path
import me.zhanghai.android.fastscroll.PopupTextProvider
import me.zhanghai.android.files.R
import me.zhanghai.android.files.coil.AppIconPackageName
import me.zhanghai.android.files.databinding.FileItemBinding
import me.zhanghai.android.files.file.FileItem
import me.zhanghai.android.files.file.fileSize
import me.zhanghai.android.files.file.formatShort
import me.zhanghai.android.files.file.iconRes
import me.zhanghai.android.files.provider.archive.isArchivePath
import me.zhanghai.android.files.settings.Settings
import me.zhanghai.android.files.ui.AnimatedListAdapter
import me.zhanghai.android.files.ui.CheckableItemBackground
import me.zhanghai.android.files.util.layoutInflater
import me.zhanghai.android.files.util.valueCompat
import java.util.Locale

class FileListAdapter(
    private val listener: Listener
) : AnimatedListAdapter<FileItem, FileListAdapter.ViewHolder>(CALLBACK), PopupTextProvider {
    private var isSearching = false

    private lateinit var _comparator: Comparator<FileItem>
    var comparator: Comparator<FileItem>
        get() = _comparator
        set(value) {
            _comparator = value
            if (!isSearching) {
                super.replace(list.sortedWith(value), true)
                rebuildFilePositionMap()
            }
        }

    var pickOptions: PickOptions? = null
        set(value) {
            field = value
            notifyItemRangeChanged(0, itemCount, PAYLOAD_STATE_CHANGED)
        }

    private val selectedFiles = fileItemSetOf()

    private val filePositionMap = mutableMapOf<Path, Int>()

    private lateinit var _nameEllipsize: TextUtils.TruncateAt
    var nameEllipsize: TextUtils.TruncateAt
        get() = _nameEllipsize
        set(value) {
            _nameEllipsize = value
            notifyItemRangeChanged(0, itemCount, PAYLOAD_STATE_CHANGED)
        }

    fun replaceSelectedFiles(files: FileItemSet) {
        val changedFiles = fileItemSetOf()
        val iterator = selectedFiles.iterator()
        while (iterator.hasNext()) {
            val file = iterator.next()
            if (file !in files) {
                iterator.remove()
                changedFiles.add(file)
            }
        }
        for (file in files) {
            if (file !in selectedFiles) {
                selectedFiles.add(file)
                changedFiles.add(file)
            }
        }
        for (file in changedFiles) {
            val position = filePositionMap[file.path]
            position?.let { notifyItemChanged(it, PAYLOAD_STATE_CHANGED) }
        }
    }

    private fun selectFile(file: FileItem) {
        if (!isFileSelectable(file)) {
            return
        }
        val selected = file in selectedFiles
        val pickOptions = pickOptions
        if (!selected && pickOptions != null && !pickOptions.allowMultiple) {
            listener.clearSelectedFiles()
        }
        listener.selectFile(file, !selected)
    }

    fun selectAllFiles() {
        val files = fileItemSetOf()
        for (index in 0 until itemCount) {
            val file = getItem(index)
            if (isFileSelectable(file)) {
                files.add(file)
            }
        }
        listener.selectFiles(files, true)
    }

    private fun isFileSelectable(file: FileItem): Boolean {
        val pickOptions = pickOptions ?: return true
        return if (pickOptions.pickDirectory) {
            file.attributes.isDirectory
        } else {
            !file.attributes.isDirectory && pickOptions.mimeTypes.any { it.match(file.mimeType) }
        }
    }

    override fun clear() {
        super.clear()

        rebuildFilePositionMap()
    }

    @Deprecated("", ReplaceWith("replaceListAndSearching(list, searching)"))
    override fun replace(list: List<FileItem>, clear: Boolean) {
        throw UnsupportedOperationException()
    }

    fun replaceListAndIsSearching(list: List<FileItem>, isSearching: Boolean) {
        val clear = this.isSearching != isSearching
        this.isSearching = isSearching
        super.replace(if (!isSearching) list.sortedWith(comparator) else list, clear)
        rebuildFilePositionMap()
    }

    private fun rebuildFilePositionMap() {
        filePositionMap.clear()
        for (index in 0 until itemCount) {
            val file = getItem(index)
            filePositionMap[file.path] = index
        }
    }

    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): ViewHolder =
        ViewHolder(
            FileItemBinding.inflate(parent.context.layoutInflater, parent, false)
        ).apply {
            binding.itemLayout.background =
                CheckableItemBackground.create(binding.itemLayout.context)
            popupMenu = PopupMenu(binding.menuButton.context, binding.menuButton)
                .apply { inflate(R.menu.file_item) }
            binding.menuButton.setOnClickListener { popupMenu.show() }
        }

    override fun onBindViewHolder(holder: ViewHolder, position: Int) {
        throw UnsupportedOperationException()
    }

    override fun onBindViewHolder(holder: ViewHolder, position: Int, payloads: List<Any>) {
        val file = getItem(position)
        val binding = holder.binding
        val isDirectory = file.attributes.isDirectory
        val enabled = isFileSelectable(file) || isDirectory
        binding.itemLayout.isEnabled = enabled
        binding.menuButton.isEnabled = enabled
        val menu = holder.popupMenu.menu
        val path = file.path
        val hasPickOptions = pickOptions != null
        val isReadOnly = path.fileSystem.isReadOnly
        menu.findItem(R.id.action_cut).isVisible = !hasPickOptions && !isReadOnly
        menu.findItem(R.id.action_copy).isVisible = !hasPickOptions
        val checked = file in selectedFiles
        binding.itemLayout.isChecked = checked
        val nameEllipsize = nameEllipsize
        binding.nameText.ellipsize = nameEllipsize
        binding.nameText.isSelected = nameEllipsize == TextUtils.TruncateAt.MARQUEE
        if (payloads.isNotEmpty()) {
            return
        }
        bindViewHolderAnimation(holder)
        binding.itemLayout.setOnClickListener {
            if (selectedFiles.isEmpty()) {
                listener.openFile(file)
            } else {
                selectFile(file)
            }
        }
        binding.itemLayout.setOnLongClickListener {
            if (selectedFiles.isEmpty()) {
                selectFile(file)
            } else {
                listener.openFile(file)
            }
            true
        }
        binding.iconLayout.setOnClickListener { selectFile(file) }
        binding.iconImage.setImageResource(file.mimeType.iconRes)
        binding.iconImage.isVisible = true
        binding.thumbnailImage.clear()
        binding.thumbnailImage.setImageDrawable(null)
        val supportsThumbnail = file.supportsThumbnail
        binding.thumbnailImage.isVisible = supportsThumbnail
        val attributes = file.attributes
        if (supportsThumbnail) {
            binding.thumbnailImage.loadAny(path to attributes) {
                listener { _, _ -> binding.iconImage.isVisible = false }
            }
        }
        binding.appIconBadgeImage.clear()
        binding.appIconBadgeImage.setImageDrawable(null)
        val appDirectoryPackageName = file.appDirectoryPackageName
        val hasAppIconBadge = appDirectoryPackageName != null
        binding.appIconBadgeImage.isVisible = hasAppIconBadge
        if (hasAppIconBadge) {
            binding.appIconBadgeImage.loadAny(AppIconPackageName(appDirectoryPackageName!!))
        }
        val badgeIconRes = if (file.attributesNoFollowLinks.isSymbolicLink) {
            if (file.isSymbolicLinkBroken) {
                R.drawable.error_badge_icon_18dp
            } else {
                R.drawable.symbolic_link_badge_icon_18dp
            }
        } else {
            null
        }
        val hasBadge = badgeIconRes != null
        binding.badgeImage.isVisible = hasBadge
        if (hasBadge) {
            binding.badgeImage.setImageResource(badgeIconRes!!)
        }
        binding.nameText.text = file.name
        binding.descriptionText.text = if (isDirectory) {
            null
        } else {
            val context = binding.descriptionText.context
            val lastModificationTime = attributes.lastModifiedTime().toInstant()
                .formatShort(context)
            val size = attributes.fileSize.formatHumanReadable(context)
            val descriptionSeparator = context.getString(R.string.file_item_description_separator)
            listOf(lastModificationTime, size).joinToString(descriptionSeparator)
        }
        val isArchivePath = path.isArchivePath
        menu.findItem(R.id.action_copy)
            .setTitle(if (isArchivePath) R.string.file_item_action_extract else R.string.copy)
        menu.findItem(R.id.action_delete).isVisible = !isReadOnly
        menu.findItem(R.id.action_rename).isVisible = !isReadOnly
        menu.findItem(R.id.action_extract).isVisible = file.isArchiveFile
        menu.findItem(R.id.action_archive).isVisible = !isArchivePath
        menu.findItem(R.id.action_add_bookmark).isVisible = isDirectory
        holder.popupMenu.setOnMenuItemClickListener {
            when (it.itemId) {
                R.id.action_open_with -> {
                    listener.openFileWith(file)
                    true
                }
                R.id.action_cut -> {
                    listener.cutFile(file)
                    true
                }
                R.id.action_copy -> {
                    listener.copyFile(file)
                    true
                }
                R.id.action_delete -> {
                    listener.confirmDeleteFile(file)
                    true
                }
                R.id.action_rename -> {
                    listener.showRenameFileDialog(file)
                    true
                }
                R.id.action_extract -> {
                    listener.extractFile(file)
                    true
                }
                R.id.action_archive -> {
                    listener.showCreateArchiveDialog(file)
                    true
                }
                R.id.action_share -> {
                    listener.shareFile(file)
                    true
                }
                R.id.action_copy_path -> {
                    listener.copyPath(file)
                    true
                }
                R.id.action_add_bookmark -> {
                    listener.addBookmark(file)
                    true
                }
                R.id.action_create_shortcut -> {
                    listener.createShortcut(file)
                    true
                }
                R.id.action_properties -> {
                    listener.showPropertiesDialog(file)
                    true
                }
                else -> false
            }
        }
    }

    override fun getPopupText(position: Int): String {
        val file = getItem(position)
        return file.name.take(1).uppercase(Locale.getDefault())
    }

    override val isAnimationEnabled: Boolean
        get() = Settings.FILE_LIST_ANIMATION.valueCompat

    companion object {
        private val PAYLOAD_STATE_CHANGED = Any()

        private val CALLBACK = object : DiffUtil.ItemCallback<FileItem>() {
            override fun areItemsTheSame(oldItem: FileItem, newItem: FileItem): Boolean =
                oldItem.path == newItem.path

            override fun areContentsTheSame(oldItem: FileItem, newItem: FileItem): Boolean =
                oldItem == newItem
        }
    }

    class ViewHolder(val binding: FileItemBinding) : RecyclerView.ViewHolder(binding.root) {
        lateinit var popupMenu: PopupMenu
    }

    interface Listener {
        fun clearSelectedFiles()
        fun selectFile(file: FileItem, selected: Boolean)
        fun selectFiles(files: FileItemSet, selected: Boolean)
        fun openFile(file: FileItem)
        fun openFileWith(file: FileItem)
        fun cutFile(file: FileItem)
        fun copyFile(file: FileItem)
        fun confirmDeleteFile(file: FileItem)
        fun showRenameFileDialog(file: FileItem)
        fun extractFile(file: FileItem)
        fun showCreateArchiveDialog(file: FileItem)
        fun shareFile(file: FileItem)
        fun copyPath(file: FileItem)
        fun addBookmark(file: FileItem)
        fun createShortcut(file: FileItem)
        fun showPropertiesDialog(file: FileItem)
    }
}
